v3/src/models/data/data-set.ts

```diff
@@ -919,22 +919,74 @@ export const DataSet = V2Model.named("DataSet").props({
       },
 
       removeCases(caseIDs: string[]) {
-        caseIDs.forEach((caseID) => {
-          const index = self.getItemIndex(caseID)
+        // Remove the items last -> first, so we only have to update itemInfo once
+        const items = caseIDs.map(id => ({ id, index: self.getItemIndex(id) }))
+          .filter(info => info.index != null) as { id: string, index: number }[]
+        items.sort((a, b) => b.index - a.index)
+        const firstIndex = items[items.length - 1]?.index ?? -1
+        items.forEach(({ id: caseID, index }) => {
           if (index != null) {
             self.itemIds.splice(index, 1)
             self.attributes.forEach((attr) => {
               attr.removeValues(index)
             })
             self.selection.delete(caseID)
             self.itemInfoMap.delete(caseID)
-            for (let i = index; i < self.items.length; ++i) {
-              const itemId = self.items[i].__id__
-              const itemInfo = self.itemInfoMap.get(itemId)
-              if (itemInfo) itemInfo.index = i
-            }
           }
         })
+        if (firstIndex >= 0) {
+          for (let i = firstIndex; i < self.itemIds.length; ++i) {
+            const itemId = self.itemIds[i]
+            const itemInfo = self.itemInfoMap.get(itemId)
+            if (itemInfo) itemInfo.index = i
+          }
+        }
+      },
+
+      // If order is "first", items will be moved to the front. Otherwise, they are moved to the end.
+      moveItems(itemIds: string[], order?: string) {
+        const first = order === "first"
+        const indices = itemIds.map(itemId => self.getItemIndex(itemId)).filter(index => index != null)
+          .sort((a: number, b: number) => b - a) // Reverse order
+        const items = indices.map(index => {
+          const item = { index, item: self.items[index], values: [] as { strValue: string, numValue: number }[] }
+          self.attributes.forEach(attr => item.values.push({
+            strValue: attr.strValues[index],
+            numValue: attr.numValues[index]
+          }))
+          return item
+        })
+
+        // Remove from ordered arrays
+        items.forEach(({ index }) => {
+          self.itemIds.splice(index, 1)
+          self.attributes.forEach(attr => attr.removeValues(index))
+        })
+
+        // Add back to ordered arrays
+        if (!first) items.reverse()
+        items.forEach(({ item, values }) => {
+          if (first) {
+            self.itemIds.unshift(item.__id__)
+            self.attributes.forEach((attr, index) => {
+              attr.strValues.unshift(values[index].strValue)
+              attr.numValues.unshift(values[index].numValue)
+            })
+          } else {
+            self.itemIds.push(item.__id__)
+            self.attributes.forEach((attr, index) => {
+              attr.strValues.push(values[index].strValue)
+              attr.numValues.push(values[index].numValue)
+            })
+          }
+        })
```

Comment:
Seems like you should be able to use `splice()` to insert all of the cases at once rather than having to loop through and add them one at a time. If you make the API change I suggested above, then there would be a `beforeIndex` determined from the `beforeId` and then each array can be updated with `array.splice(beforeIndex, 0, ...movedValues)`.